lib/wibox/layout/manual.lua

```diff
@@ -0,0 +1,211 @@
+end
+
+function manual_layout:layout(context, width, height)
+    local res = {}
+
+    for k, v in ipairs(self._private.widgets) do
+        local pt = self._private.pos[k] or {x=0,y=0}
+        local w, h = base.fit_widget(self, context, v, width, height)
+
+        -- Make sure the signature is compatible with `awful.placement`. `Wibox`,
+        -- doesn't depend on `awful`, but it is still nice not to have to code
+        -- geometry functions again and again.
+        if type(pt) == "function" or (getmetatable(pt) or {}).__call then
+            local geo = {
+                x      = 0,
+                y      = 0,
+                width  = w,
+                height = h,
+                geometry = geometry,
+            }
+            pt = pt(geo, {
+                parent = {
+                    x=0, y=0, width = width, height = height, geometry = geometry
+                }
+            })
+            -- Trick to ensure compatibility with `awful.placement`
+            gtable.crush(pt, geo._new_geo or {})
+        end
+
+        assert(pt.x)
+        assert(pt.y)
+        assert(pt.width or w)
+        assert(pt.height or h)
+
+        table.insert(res, base.place_widget_at(
+            v, pt.x, pt.y, pt.width or w, pt.height or h
+        ))
+    end
+
+    return res
+end
+
+function manual_layout:add(...)
+    local wdgs = {...}
+    local old_count = #self._private.widgets
+    gtable.merge(self._private.widgets, {...})
+
+    -- Add the points
+    for k, v in ipairs(wdgs) do
+        if v.point then
+            self._private.pos[old_count+k] = v.point
+        end
+    end
+
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Add a widget at a specific point.
+--
+-- The point can either be a function or a table. The table follow the generic
+-- geometry format used elsewhere in Awesome.
+--
+-- * *x*: The horizontal position.
+-- * *y*: The vertical position.
+-- * *width*: The width.
+-- * *height*: The height.
+--
+-- If a function is used, it follows the same prototype as `awful.placement`
+-- functions.
+--
+-- * *geo*:
+--     * *x*: The horizontal position (always 0).
+--     * *y*: The vertical position (always 0).
+--     * *width*: The width.
+--     * *height*: The height.
+--     * *geometry*: A function to get or set the geometry (for compatibility)
+-- * *args*:
+--     * *parent* The layout own geometry
+--         * *x*: The horizontal position (always 0).
+--         * *y*: The vertical position (always 0).
+--         * *width*: The width.
+--         * *height*: The height.
+--         * *geometry*: A function to get or set the geometry (for compatibility)
+--
+--@DOC_wibox_layout_manual_add_at_EXAMPLE@
+-- @tparam widget widget The widget.
+-- @tparam table|function point Either an `{x=x,y=y}` table or a function
+--  returning the new geometry.
+function manual_layout:add_at(widget, point)
+    self._private.pos[#self._private.widgets+1] = point
+    self:add(widget)
```

Comment:
Also, should `:add` and `:add_at` somehow verify the point that they got? E.g. if it is a function, call it with some dummy values to see that it is well-behaved and otherwise throw an error. That's better than the whole widget not being drawn because of an error that made `:layout()` return by error.